In [ ]:
# Author : Arewa Morountudun Ojelade
# Date : 7/1/2025
# Organization : Miami Machine Learning
# Sponsor : Neural Transformers

# Neural Network Vampire Number Classification with Sigmoid Function

In [ ]:
import numpy as np
import pandas as pd
import math
import random
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


Here we are populating the interval in a numpy array

In [ ]:
interval = np.arange(100000,999999);
m = interval.shape[0]

A series of lambda functions to perform various tasks required in fang  identification

In [ ]:
# These lambda functions return an array
summation = np.vectorize(lambda x: sum(int(digit) for digit in str(x)))    # sums up digits in a value for each element in an array
count = np.vectorize(lambda x: sum(1 for digit in str(x)))                 # counts number of digits in a value for each element in an array
even = np.vectorize(lambda x: len(str(x)) % 2 == 0)                        # qualifies each value as even or odd for each element in an array
natural = np.vectorize(lambda x: int(str(x).isdigit and x >= 0))           # qualifies each value as a natural number for each element in an arry

# These lambda functions return a boolean
trailing = lambda x, y : not(str(x)[-1] == "0" and  str(y)[-1] == "0")                           # boolean of trailing feature
permute = lambda x, y : sorted(list(str(x))) == sorted(list(str(y)))                             # boolean of permutation feature
half_as_many = lambda x, y, z : len(x) == int(len(str(z))/2)  and len(y) == int(len(str(z))/2)   # boolean of half as many feature

Declaring all array variables necessary for creating a data set and populating with zeros

In [ ]:

fang_1 = np.zeros(m)
fang_2 = np.zeros(m)
fang_even = np.zeros(m)
fang_half_as_many =  np.zeros(m)
fang_permutation =  np.zeros(m)
fang_trailing_zeros =  np.zeros(m)
vampire_number =  np.zeros(m)

vampire_count = [0]  # declaring vampiring number count at 0


In [ ]:

# call to verify feature lambda functions
print(trailing('260','401'))
print(permute("201510" , "102510"))
print(trailing(201,510) and permute("201510" , "102510"))


True
True
True


In [ ]:
def getFangs(value, sum, index):

  #interval_limit = [9, 13, 18, 22, 27, 31, 36, 40]
  #if sum in interval_limit:
    #continue

  # declaring feature variables
  trailing_feature = 0
  permutation_feature = 0
  half_as_many_feature = 0

  for i in range(2, int(math.sqrt(value) + 1)):
    # iterating through to find factors
    if int(value) % i == 0 :
      fang1 = str(i)
      fang2 = str(int(value/i))

      # validating features for each factor
      trailing_feature = int(trailing(fang1, fang2))
      permutation_feature = int(permute(fang1+fang2, str(value)))
      half_as_many_feature = int(half_as_many(fang1, fang2, value))

      # if factors are validated as fangs
      if trailing_feature and permutation_feature and half_as_many_feature:
        #classify as vampire number
        vampire_number[index] += 1
        fang_trailing_zeros[index] += trailing_feature
        fang_half_as_many[index] += half_as_many_feature
        fang_permutation[index] += permutation_feature

        # populate fang arrays with associated fang values
        fang_1[index] += int(fang1)
        fang_2[index] += int(fang2)

        # print to console
        print(fang1, fang2, value)
        print("features : " ,trailing_feature, half_as_many_feature, permutation_feature)

        #increment vampire count
        vampire_count[0] += 1
        return 1
  return 0

In [ ]:
natural_number = natural(interval)      # identifying natural number boolean of values in interval
digit_count = count(interval)           # identifying digit count of values in interval
even_number_of_digits = even(interval)  # identifying number of digits of values interval
digit_sum = summation(interval)         # identifying digit sum of values in interval

# identifying fangs and classifying features of each value in interval
pair_of_fangs = [getFangs(x, y,z) for x,y,z in zip(interval,digit_sum,range(m))]

201 510 102510
features :  1 1 1
260 401 104260
features :  1 1 1
210 501 105210
features :  1 1 1
204 516 105264
features :  1 1 1
150 705 105750
features :  1 1 1
135 801 108135
features :  1 1 1
158 701 110758
features :  1 1 1
152 761 115672
features :  1 1 1
161 725 116725
features :  1 1 1
167 701 117067
features :  1 1 1
141 840 118440
features :  1 1 1
201 600 120600
features :  1 1 1
231 534 123354
features :  1 1 1
281 443 124483
features :  1 1 1
152 824 125248
features :  1 1 1
231 543 125433
features :  1 1 1
204 615 125460
features :  1 1 1
251 500 125500
features :  1 1 1
201 627 126027
features :  1 1 1
261 486 126846
features :  1 1 1
140 926 129640
features :  1 1 1
179 725 129775
features :  1 1 1
311 422 131242
features :  1 1 1
323 410 132430
features :  1 1 1
315 423 133245
features :  1 1 1
317 425 134725
features :  1 1 1
231 588 135828
features :  1 1 1
351 387 135837
features :  1 1 1
215 635 136525
features :  1 1 1
146 938 136948
features :  1 1 1
350 401 14

In [ ]:
# print to console how many numbers were found
print(f"{vampire_count[0]} Vampire Numbers Found")


148 Vampire Numbers Found


In [ ]:
# zipping the list of attributes to create a data set
data = list(zip(interval.tolist(),
                digit_sum.tolist(),
                natural_number.tolist(),
                even_number_of_digits.tolist(),
                fang_half_as_many.tolist(),
                fang_permutation.tolist(),
                fang_trailing_zeros.tolist(),
                fang_1.tolist(),
                fang_2.tolist(),
                vampire_number.tolist()))



In [ ]:
from dataclasses import make_dataclass
headers = ["number", "digit_sum", "natural_number", "number_even_digits",
           "half_as_many_feature",
           "permutation_feature", "trailing_zeros_feature","fang_1", "fang_2",
           "vampire_number"]

# creating a pandas dataframe in order to save to csv
df = pd.DataFrame(np.array(data), columns = headers)
df.head()

number  digit_sum  natural_number  number_even_digits  \
0  100000.0        1.0             1.0                 1.0   
1  100001.0        2.0             1.0                 1.0   
2  100002.0        3.0             1.0                 1.0   
3  100003.0        4.0             1.0                 1.0   
4  100004.0        5.0             1.0                 1.0   

   half_as_many_feature  permutation_feature  trailing_zeros_feature  fang_1  \
0                   0.0                  0.0                     0.0     0.0   
1                   0.0                  0.0                     0.0     0.0   
2                   0.0                  0.0                     0.0     0.0   
3                   0.0                  0.0                     0.0     0.0   
4                   0.0                  0.0                     0.0     0.0   

   fang_2  vampire_number  
0     0.0             0.0  
1     0.0             0.0  
2     0.0             0.0  
3     0.0             0.0  
4     0.0             0.0

In [ ]:
# changing data type of dataframe to integer
df = df.astype({"number": "int", "digit_sum" : "int","natural_number": "int",
                "number_even_digits" : "int",
                "half_as_many_feature": "int","permutation_feature": "int",
                "trailing_zeros_feature": "int","fang_1": "int", "fang_2" : "int",
                "vampire_number": "int"})
df.head()


number  digit_sum  natural_number  number_even_digits  \
0  100000          1               1                   1   
1  100001          2               1                   1   
2  100002          3               1                   1   
3  100003          4               1                   1   
4  100004          5               1                   1   

   half_as_many_feature  permutation_feature  trailing_zeros_feature  fang_1  \
0                     0                    0                       0       0   
1                     0                    0                       0       0   
2                     0                    0                       0       0   
3                     0                    0                       0       0   
4                     0                    0                       0       0   

   fang_2  vampire_number  
0       0               0  
1       0               0  
2       0               0  
3       0               0  
4       0               0

In [ ]:
# Taking a look at all rows with identified vampire numbers. Verifying length
vamp = df.loc[df["vampire_number"] == 1]
vamp = pd.DataFrame(np.array(vamp), columns = headers)

In [ ]:
print(vamp['number'])

0      102510
1      104260
2      105210
3      105264
4      105750
        ...  
143    809964
144    815958
145    829696
146    841995
147    939658
Name: number, Length: 148, dtype: int64


In [ ]:
#Saving dataframe to csv file
df.to_csv("vampire.csv", index = False)

In [ ]:
df.to_json("vampire.json", orient = "records", indent = 4)

In [ ]:
df.iloc[4260]

number                    104260
digit_sum                     13
natural_number                 1
number_even_digits             1
half_as_many_feature           1
permutation_feature            1
trailing_zeros_feature         1
fang_1                       260
fang_2                       401
vampire_number                 1
Name: 4260, dtype: int64

In [ ]:
df = pd.read_csv("vampire.csv")
df.head()

number  digit_sum  natural_number  number_even_digits  \
0  100000          1               1                   1   
1  100001          2               1                   1   
2  100002          3               1                   1   
3  100003          4               1                   1   
4  100004          5               1                   1   

   half_as_many_feature  permutation_feature  trailing_zeros_feature  fang_1  \
0                     0                    0                       0       0   
1                     0                    0                       0       0   
2                     0                    0                       0       0   
3                     0                    0                       0       0   
4                     0                    0                       0       0   

   fang_2  vampire_number  
0       0               0  
1       0               0  
2       0               0  
3       0               0  
4       0               0

Here we drop redundant columns, or columns to closely associated with the output label.

In [ ]:
x = df.drop(columns = ["natural_number","number_even_digits", "fang_1", "fang_2"])
y = df["vampire_number"]


Here we balance our dataset since there are only 148 vampire numbers in order to have sufficient training data.

In [ ]:
x1 = x.loc[x["vampire_number"] == 1]
x0 = x.loc[x["vampire_number"] == 0]

x1 = np.array(x1)
x0 = np.array(x0)

x1 = np.tile(x1,(6000, 1))

print(x1.shape)
print(x0.shape)


(888000, 6)
(899851, 6)


Here we separate the features from the output label.

In [ ]:
print(x1[0])
x = list(x1[:,: 5]) + list(x0[:,: 5])
y = list(x1[:,-1]) + list(x0[:,-1])
x = np.array(x)
y = np.array(y)
print(x[0])

[102510      9      1      1      1      1]
[102510      9      1      1      1]


In [ ]:
print(x.shape)
print(y.shape)

(1787851, 5)
(1787851,)


Once our dataset has been generated we can start splitting our data into training, and test sets.

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y,test_size=.2, random_state = 42)

In [ ]:
x_num_train = x_train
x_num_test = x_test
x_train = x_train[:,1:5]
x_test = x_test[:,1:5]

print(x_num_train[0])
print(x_num_test[0])
print(x_train[0])
print(x_test[0])

[148406     23      0      0      0]
[780636     30      0      0      0]
[23  0  0  0]
[30  0  0  0]


In [ ]:
model = Sequential(
    [
        tf.keras.Input(shape=(4,)),    #specify input size

        Dense(4, activation = 'sigmoid', name ='dense'),
        Dense(1, activation = 'sigmoid', name ='dense_1'),
    ]
)


In [ ]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 4)              │            20 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │             5 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 25 (100.00 B)

 Trainable params: 25 (100.00 B)

 Non-trainable params: 0 (0.00 B)

In [ ]:
[layer1,layer2] = model.layers
W1,b1 = layer1.get_weights()
W2,b2 = layer2.get_weights()
print(f"W1{W1.shape}:\n", W1, f"\nb1{b1.shape}:", b1)
print(f"W2{W2.shape}:\n", W2, f"\nb2{b2.shape}:", b2)

W1(4, 4):
 [[-0.46249196  0.40040153  0.10136181  0.01477814]
 [-0.1278544   0.05001938  0.18156022  0.48989755]
 [-0.13933408  0.57831603  0.6031919   0.6287605 ]
 [ 0.48386377  0.79071826 -0.22402233  0.52295715]] 
b1(4,): [0. 0. 0. 0.]
W2(4, 1):
 [[-0.9074019 ]
 [-0.7676741 ]
 [ 0.6712388 ]
 [-0.16990238]] 
b2(1,): [0.]


Here we are normalizing our dataset in order to improve model performance.
Normalization is a technique to improve data integrity.

In [ ]:
norm_l = tf.keras.layers.Normalization(axis=-1)
x_train = np.array(x_train)
norm_l.adapt(x_train)  # learns mean, variance
Xn = norm_l(x_train)


In [ ]:
model.compile(
    loss = tf.keras.losses.BinaryCrossentropy(),
    optimizer = tf.keras.optimizers.Adam(learning_rate=0.01),
)

model.fit(
    Xn,y_train,
    epochs=10,
)

Epoch 1/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 76s 2ms/step - loss: 0.0092
Epoch 2/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 81s 2ms/step - loss: 1.4955e-08
Epoch 3/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 82s 2ms/step - loss: 1.4814e-08
Epoch 4/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 83s 2ms/step - loss: 1.4806e-08
Epoch 5/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 81s 2ms/step - loss: 1.4812e-08
Epoch 6/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 82s 2ms/step - loss: 1.4792e-08
Epoch 7/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 75s 2ms/step - loss: 1.4795e-08
Epoch 8/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 75s 2ms/step - loss: 1.4822e-08
Epoch 9/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 80s 2ms/step - loss: 1.4809e-08
Epoch 10/10
44697/44697 ━━━━━━━━━━━━━━━━━━━━ 75s 2ms/step - loss: 1.4789e-08


In [ ]:
[layer1,layer2] = model.layers
W1,b1 = layer1.get_weights()
W2,b2 = layer2.get_weights()
print(f"W1{W1.shape}:\n", W1, f"\nb1{b1.shape}:", b1)
print(f"W2{W2.shape}:\n", W2, f"\nb2{b2.shape}:", b2)

W1(4, 4):
 [[-0.01440327  0.01158994 -0.04043884 -0.03876886]
 [-2.4231057   1.3806235   1.5378193   1.7086192 ]
 [-2.5221055   1.897896    2.5364068   2.5904531 ]
 [-1.5394851   2.8638816   2.5441923   2.5867457 ]] 
b1(4,): [ 0.44512382 -0.44303176 -0.41411808 -0.41636682]
W2(4, 1):
 [[-15.594596 ]
 [  9.89562  ]
 [  7.2674074]
 [  9.050568 ]] 
b2(1,): [-8.78962]


In [ ]:
norm_l = tf.keras.layers.Normalization(axis=-1)
x_test = np.array(x_test)
norm_l.adapt(x_test)  # learns mean, variance
Xn = norm_l(x_test)


In [ ]:
x

In [ ]:
predictions = model.predict(Xn)
print("predictions = \n", predictions)

11175/11175 ━━━━━━━━━━━━━━━━━━━━ 11s 1ms/step
predictions = 
 [[2.6842932e-11]
 [1.0000000e+00]
 [2.6848569e-11]
 ...
 [1.0000000e+00]
 [1.0000000e+00]
 [1.0000000e+00]]


In [ ]:
number1 = x_num_test[1][0]
number2 = x_num_test[2][0]
print(number1)
print(number2)

829696
601309


In [ ]:
# map predictions array to binary classification
predictions = list(map(lambda x : 0 if x < .5 else 1 ,predictions))

In [ ]:
# print number prediction and actual
print(f"  Number  Prediction  Actual" )
print(f"  {number1:6} {predictions[1]:7} {y_test[1]:8}")
print(f"  {number2:6} {predictions[2]:7} {y_test[2]:8}")



  Number  Prediction  Actual
  829696       1        1
  601309       0        0


In [ ]:
print("Accuracy:", accuracy_score(y_test, predictions))
print("Confusion Matrix:\n", confusion_matrix(y_test, predictions))
print("Classification Report:\n", classification_report(y_test, predictions))


Accuracy: 1.0
Confusion Matrix:
 [[180013      0]
 [     0 177558]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00    180013
           1       1.00      1.00      1.00    177558

    accuracy                           1.00    357571
   macro avg       1.00      1.00      1.00    357571
weighted avg       1.00      1.00      1.00    357571



Running Logistic Regression for comparison

In [ ]:
y_pred = LogisticRegression(solver='liblinear', random_state=1234).fit(x_train, y_train).predict(x_test)
print("Accuracy Reverification:", accuracy_score(y_test, y_pred))
print("Confusion Matrix Reverification:\n", confusion_matrix(y_test, y_pred))
print("Classification Report Reverification:\n", classification_report(y_test, y_pred))


Accuracy Reverification: 1.0
Confusion Matrix Reverification:
 [[180013      0]
 [     0 177558]]
Classification Report Reverification:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00    180013
           1       1.00      1.00      1.00    177558

    accuracy                           1.00    357571
   macro avg       1.00      1.00      1.00    357571
weighted avg       1.00      1.00      1.00    357571

